packages/hash/frontend/src/blocks/page/PageTitle/PageTitle.tsx

```diff
@@ -0,0 +1,122 @@
+import { Typography } from "@mui/material";
+import {
+  ChangeEventHandler,
+  FocusEventHandler,
+  KeyboardEventHandler,
+  useRef,
+  FunctionComponent,
+} from "react";
+import { useBlockProtocolUpdateEntity } from "../../../components/hooks/blockProtocolFunctions/useBlockProtocolUpdateEntity";
+import { rewriteEntityIdentifier } from "../../../lib/entities";
+import { usePageContext } from "../PageContext";
+import { cleanUpTitle, focusEditorBeginning, isValidPageTitle } from "./utils";
+
+type PageTitleProps = {
+  accountId: string;
+  entityId: string;
+  value: string;
+};
+
+// TODO: Add read-only mode based on page permissions
+export const PageTitle: FunctionComponent<PageTitleProps> = ({
+  accountId,
+  entityId,
+  value,
+}) => {
+  // TODO: Display update error once expected UX is discussed
+  const { updateEntity, updateEntityLoading } =
+    useBlockProtocolUpdateEntity(true);
+  const { editorView } = usePageContext();
+
+  const titleValueRef = useRef(value);
+
+  const handleInputChange: ChangeEventHandler<HTMLHeadingElement> = (event) => {
+    const { currentTarget } = event;
+    titleValueRef.current = currentTarget.textContent || "";
+
+    /**
+     * this prevents two issues
+     * 1 - when user pastes styled stuff into title
+     * other way of achieving this is implementing https://htmldom.dev/paste-as-plain-text/
+     * this looks like working well for now on all browser.
+     * we can switch to paste-as-plain-text later if we decide this solution is not enough
+     * 2 - firefox & safari renders a `br` tag when user clears the editable area,
+     * which prevents css selector `:empty` from working. This removes the `br` as well
+     */
+    if (currentTarget?.children.length) {
+      currentTarget.innerText = currentTarget.textContent || "";
+    }
+  };
+
+  const handleInputKeyDown: KeyboardEventHandler<HTMLInputElement> = (
+    event,
+  ) => {
+    if (event.key === "Enter" || event.key === "Escape") {
+      event.currentTarget.blur();
+    }
+
+    if (event.key === "ArrowDown") {
+      const { anchorOffset, type } = window.getSelection() || {};
+
+      const isCaret = type === "Caret";
+      const isCaretAtEnd = anchorOffset === titleValueRef?.current.length;
```

Comment:
This doesn't quite work as expected. You wouldn't expect a down arrow key to jump you to the end of a line if there's text below it. I think the down arrow key should allows jump to the editor if its on the last line (as the title could wrap). Similarly, when there's multiple lines and you're focused in the editor and press up, the cursor jumps to the start of the title, not where you would expect it. Additionally the left/right arrow keys don't work.

I reckon the *best* and most natural way to handle this would be for the title to be part of the prosemirror contenteditable so that cursors are handled as expected. It might have been worth discussing the approach for this before work was put into this approach. Doing it with prosemirror means the cursor will jump to the correct position.

Short of that, we could make quick wins on the UX – i.e, when you press up from the editor, it the cursor goes to the last line of the title, and when you press down on the last line, it jumps to the editor. This may not be worth it though unless we want to go the whole way of moving the title into contenteditable.

Will leave to @nonparibus and @jtewright to decide.

See loom below.  https://www.loom.com/share/4a823a675f69433482d967ccdc71bebe

Reply:
I don't see why the title shouldn't be handled by the text editor the same way as any other text (albeit as a block that cannot be removed from the page). Doing so seems sensible to me.

Reply:
I definitely think that title should be handled by text editor, because It's definitely not worth it to pursue e flawless UX by handling cursor placement / focus switch between editor and a separate title entity on the page. But I think `moving header into Editor` is out-of-scope from this PR.